src/sources/apache_metrics/parser.rs

```diff
@@ -0,0 +1,837 @@
+use crate::event::metric::{Metric, MetricKind, MetricValue};
+use chrono::{DateTime, Utc};
+use std::collections::BTreeMap;
+use std::collections::HashMap;
+use std::error::Error;
+use std::fmt;
+
+/// Parses the text output from Apache's mod_status and returns:
+///
+/// - A list of metrics generated from the output
+/// - A list of parse errors that were encountered
+///
+/// # Arguments
+///
+/// - `payload` - the mod_status output
+/// - `namespace` - the namespace to put the generated metrics in
+/// - `now` - the time the payload was fetched
+/// - `tags` - any base tags to apply to the metrics
+pub fn parse(
+    payload: &str,
+    namespace: &str,
+    now: DateTime<Utc>,
+    tags: Option<&BTreeMap<String, String>>,
+) -> (Vec<Metric>, Vec<ParseError>) {
+    let mut parsed = payload
+        .lines()
+        .into_iter()
+        .filter_map(|l| {
+            let mut parts = l.splitn(2, ":");
+            let key = parts.next();
+            let value = parts.next().map(|s| s.trim());
+            match (key, value) {
+                (Some(k), Some(v)) => Some((k, v)),
+                _ => None,
+            }
+        })
+        .collect::<Vec<_>>();
+
+    // mod_status has BusyWorkers/IdleWorkers repeated
+    // TODO better way to do this without .collect()ing? Do we care?
```

Comment:
Thoughts here? I did this to dedupe the lines which required sorting.

Reply:
I'm not quite sure what the purpose of this de-duplication is.

Is it our job to filter out those duplicate status events, if it's what Apache is reporting? What happens if I call this function twice, it would filter out duplicates for each individual call, but I would still end up with two similar events since it doesn't filter duplicates across calls, right?

Reply:
I'll try to improve the comment there, but basically there is a bug in mod_status where, in the text output, `IdleWorkers` and `BusyWorkers` appear twice. If I don't dedupe we end up publishing duplicate metrics, which is probably ok, but seems a little wonky.

Apache bug: https://bz.apache.org/bugzilla/show_bug.cgi?id=63300

Reply:
Ah, right 👍

I guess one way to avoid having to collect is to keep track of the field through a boolean, and return `None` in the `filter_map` above as soon as we've flipped the boolean to `true`?

Reply:
:+1: This made me realize I could also just use a `HashMap` which would take care of making it unique: 8e3d7539 . I'm curious what you think of that.